### ISSUE TITLE
IDISP007 false positive in unit test?

### ISSUE THREAD

**Issue by Bosch-Eli-Black** [DotNetAnalyzers/IDisposableAnalyzers#353]
IDisposableAnalyzers NuGet package: 4.0.1
Visual Studio: VS2022

Code:
```c#
public abstract class AsyncExampleUnitTestBase : ExampleUnitTestBase
{
    protected static IMessageService _messageService;

    public override void TestInitialize()
    {
        base.TestInitialize();

        // I'm getting IDISP007: "Don't dispose injected" on this line
        _messageService?.Dispose();
        // If I delete the disposing statement, then I get IDISP003: "Dispose previous before re-assigning" on this line
        _messageService = TestUtils.MakeMessageService(_logger);
    }
}
```

This feels like a false positive to me, but I'm not sure what the correct behaviour should be.

Please let me know if there's any more info that I can provide 🙂

**Comment by JohanLarsson** [comment]
Ew, regression, thanks for reporting!

**Comment by JohanLarsson** [comment]
Why are you using a static field? In general the analyzer is not happy about disposing static members as it is not clear who owns it.

**Comment by Bosch-Eli-Black** [comment]
Thanks for the reply 🙂

A more full example of our current code:

```c#
public abstract class UnitTestBase {
    protected static IMessageService _messageService;

    public override void TestInitialize()
    {
        // IDISP007: "Don't dispose injected" on this line
        _messageService?.Dispose();
        // If I delete the disposing statement, then I get IDISP003: "Dispose previous before re-assigning" on this line
        _messageService = TestUtils.MakeMessageService();
    }
}

[TestClass]
public class MyUnitTest : UnitTestBase {
    [TestInitialize]
    public override void TestInitialize()
    {
        base.TestInitialize();
        
        ...
    }
}
```

I think what happened was that the `_messageService` field originally wasn't static:
```c#
public abstract class UnitTestBase {
    // Warning IDISP008: Don't assign member with injected and created disposables
    protected IMessageService _messageService;
    
    public override void TestInitialize()
    {
        // No warning
        _messageService?.Dispose();
        // No warning
        _messageService = TestUtils.MakeMessageService();
    }
}
```

Someone tried to fix the `IDISP008` warning by making the field `static`:
```c#
    // No warning!
    protected static IMessageService _messageService;

    public override void TestInitialize()
    {
        // Maybe in an earlier version of IDisposableAnalyzers there was no IDISP007 warning for these lines?
        _messageService?.Dispose();
        _messageService = TestUtils.MakeMessageService();
    }
```

This got rid of the `IDISP008` warning, earlier versions of `IDisposableAnalyzers` didn't generate the `IDISP007` warning on `_messageService?.Dispose()` (I think), and then we updated to a newer analyzer version that did generate the `IDISP007` warning 🙂

Perhaps the correct fix is to make `_messageService` be `protected IMessageService _messageService { get; private set; }` and for `UnitTestBase` to implement `IDisposable`:

```c#
public abstract class UnitTestBase : IDisposable {
    protected IMessageService _messageService { get; private set; }
    
    public override void TestInitialize()
    {
        _messageService?.Dispose();
        _messageService = TestUtils.MakeMessageService();
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (disposing)
            {
                _messageService?.Dispose();
                _onePanel?.Dispose();
            }

            _disposed = true;
        }
    }
    
    public void Dispose()
    {
        // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
        Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }
}
```

Or, better yet, use some sort of mocking library, but we haven't gotten to that yet 🙂

Not sure how helpful that ended up being, haha. Perhaps it'd make sense to have a "Don't make disposable, injected fields be `static`" analyzer? 🙂

**Comment by JohanLarsson** [comment]
> Not sure how helpful that ended up being, haha. Perhaps it'd make sense to have a "Don't make disposable, injected fields be static" analyzer? 

The wording don't dispose injected in analyzers is not awesome for static. Requires squinting to see that the value is ~injected~ into the member.

Also you are aware of #pragma right?

```cs
#pragma warning disable IDISP008 // We want it like this here because... 
```

Explicit pragma can help clarify intent even if they are not exactly pretty
